packages/blade/src/components/Table/types.ts

```diff
@@ -295,6 +292,56 @@ type TablePaginationProps = {
   showLabel?: boolean;
 };
 
+type TablePaginationType = 'client' | 'server';
+
+type TablePaginationServerProps = TablePaginationCommonProps & {
+  /**
+   * Whether the pagination is happening on client or server.
+   * If the pagination is happening on `client`, the Table component will **divide the data into pages** and show the pages based on the page size.
+   * If the pagination is happening on `server`, the Table component will **not divide the data into pages and will show all the data**. You will have to fetch data for each page as the page changes and pass it to the Table component.
+   * When paginationType is `server`, the `onPageChange` & `totalItemCount` props are required.
+   * @default 'client'
+   * */
+  paginationType?: Extract<TablePaginationType, 'server'>;
+  /**
+   * The total number of possible items in the table. This is used to calculate the total number of pages when pagination is happening on server and not all the data is fetched at once.
+   */
+  totalItemCount: number;
+  /**
+   * Callback function that is called when the page is changed
+   */
+  onPageChange: ({ page }: { page: number }) => void;
+};
+
+type TablePaginationClientProps = TablePaginationCommonProps & {
+  /**
+   * Whether the pagination is happening on client or server.
+   * If the pagination is happening on `client`, the Table component will **divide the data into pages** and show the pages based on the page size.
+   * If the pagination is happening on `server`, the Table component will **not divide the data into pages and will show all the data**. You will have to fetch data for each page as the page changes and pass it to the Table component.
+   * When paginationType is `server`, the `onPageChange` & `totalItemCount` props are required.
+   * @default 'client'
+   * */
+  paginationType?: Extract<TablePaginationType, 'client'>;
+  /**
+   * The total number of possible items in the table. This is used to calculate the total number of pages when pagination is happening on server and not all the data is fetched at once.
+   */
+  totalItemCount?: number;
+  /**
+   * Callback function that is called when the page is changed
+   */
+  onPageChange?: ({ page }: { page: number }) => void;
+};
+
+type TablePaginationProps<T> = T extends {
+  paginationType: infer TablePaginationType;
+}
+  ? TablePaginationType extends 'server'
+    ? TablePaginationServerProps
+    : TablePaginationType extends 'client'
+    ? TablePaginationClientProps
+    : T
+  : T;
```

Comment:
We don't really need to do this similar to how we did it on Text. In Text we did this because we wanted the weight prop's intellisense to not show unneeded values. But in here it doesn't matter because we are switching the prop structure. 

```diff
- type TablePaginationClientProps = TablePaginationCommonProps & {}
+ type TablePaginationClientProps = {}

+ type TablePaginationProps = (TablePaginationServerProps | TablePaginationClientProps) &
+  TablePaginationCommonProps;
```

Reply:
Basically same as how we are doing in Button, `ButtonWithoutIconProps` | `ButtonWithIconProps`

Reply:
Oh right i remember now. I will change it